src/types.rs

```diff
@@ -488,6 +500,410 @@ impl ZipFileData {
+            1u16 << 11
+        } else {
+            0
+        };
+        let encrypted_bit: u16 = if self.encrypted { 1u16 << 0 } else { 0 };
+
+        utf8_bit | encrypted_bit
+    }
+
+    fn clamp_size_field(&self, field: u64) -> u32 {
+        if self.large_file {
+            spec::ZIP64_BYTES_THR as u32
+        } else {
+            field.min(spec::ZIP64_BYTES_THR).try_into().unwrap()
+        }
+    }
+
+    pub(crate) fn local_block(&self) -> ZipResult<ZipLocalEntryBlock> {
+        let compressed_size: u32 = self.clamp_size_field(self.compressed_size);
+        let uncompressed_size: u32 = self.clamp_size_field(self.uncompressed_size);
+
+        let extra_block_len: usize = self
+            .zip64_extra_field_block()
+            .map(|block| block.full_size())
+            .unwrap_or(0);
+        let extra_field_length: u16 = (self.extra_field_len() + extra_block_len)
+            .try_into()
+            .map_err(|_| ZipError::InvalidArchive("Extra data field is too large"))?;
+
+        Ok(ZipLocalEntryBlock {
+            magic: ZipLocalEntryBlock::MAGIC,
+            version_made_by: self.version_needed(),
+            flags: self.flags(),
+            compression_method: self.compression_method.serialize_to_u16(),
+            last_mod_time: self.last_modified_time.timepart(),
+            last_mod_date: self.last_modified_time.datepart(),
+            crc32: self.crc32,
+            compressed_size,
+            uncompressed_size,
+            file_name_length: self.file_name_raw.len().try_into().unwrap(),
+            extra_field_length,
+        })
+    }
+
+    pub(crate) fn block(&self, zip64_extra_field_length: u16) -> ZipCentralEntryBlock {
+        let extra_field_len: u16 = self.extra_field_len().try_into().unwrap();
+        let central_extra_field_len: u16 = self.central_extra_field_len().try_into().unwrap();
+        ZipCentralEntryBlock {
+            magic: ZipCentralEntryBlock::MAGIC,
+            version_made_by: (self.system as u16) << 8 | (self.version_made_by as u16),
+            version_to_extract: self.version_needed(),
+            flags: self.flags(),
+            compression_method: self.compression_method.serialize_to_u16(),
+            last_mod_time: self.last_modified_time.timepart(),
+            last_mod_date: self.last_modified_time.datepart(),
+            crc32: self.crc32,
+            compressed_size: self
+                .compressed_size
+                .min(spec::ZIP64_BYTES_THR)
+                .try_into()
+                .unwrap(),
+            uncompressed_size: self
+                .uncompressed_size
+                .min(spec::ZIP64_BYTES_THR)
+                .try_into()
+                .unwrap(),
+            file_name_length: self.file_name_raw.len().try_into().unwrap(),
+            extra_field_length: zip64_extra_field_length
+                + extra_field_len
+                + central_extra_field_len,
```

Comment:
That won't work because we rely upon the precise ordering of fields in these `#[repr(packed)]` structs implementing `Block`, and `ZipLocalEntryBlock` has e.g. `version_made_by` then `flags`, while `ZipCentralEntryBlock` has `version_made_by` then `version_to_extract`. If we were to replace them, we would be parsing and serializing fields incorrectly.